### PR TITLE
Matrix assembly data

### PR DESCRIPTION
**Pull request by fritzgoebel**
This PR adds `matrix_assembly_data` which is similar to `matrix_data` but uses an unordered map to store the nonzeros and offers the option of adding values to previously inserted nonzero entries. This fuctionality is needed by openCARP during the matrix assembly.

After the assembly, to get a row major ordering of the data, one can call `matrix_assembly_data<...>.get_ordered_data()` which returns an `std::map` that can be used to read the nonzeros to a sparse matrix format.

There is a lot of code which is almost identical to `matrix_data`, with the main difference being that data is stored in an (unordered) map instead of a vector. If someone has an idea how to abstract this nicely, please share :)

### CODE REVIEW

**Comment by yhmtsai** [comment]
I guess you want to reduce the code of read, right?
the matrix read should require matrix_data in ordered.
implement the get_ordered_data() in matrix_data
(some constructors call `ensure_row_major` but some do not, so maybe need one flag to check whether running `ensure_row_major`)
The vector data can also use `for (auto &x: data.get_ordered_data())`.

Use one ReadMat inherit ReadableFromMatrixAssemblyData and ReadableFromMatrixData
ReadMat has `template<> read_impl(RangeAllowable)`
format only needs to implement `read_impl` which should be almost identical to read(matrix_assembly_data)

**Comment by upsj** [comment]
+1 to Mike's comments. We could also add a conversion from matrix_assembly_data to matrix_data to make any additional overloads unnecessary. But this would make for a slightly unintuitive interface (it's not documented that you can use matrix_assembly_data for matrix_data).
Using std::map probably introduces unnecessary overhead, since creating a vector + sorting should be faster than inserting all elements into a searchtree.

File: include/ginkgo/core/base/matrix_assembly_data.hpp [hunk 1]
```diff
@@ -0,0 +1,224 @@
...
+
+    explicit matrix_assembly_data(dim<2> size)
+        : size_{size},
+          nonzeros_(0, detail::symbolic_nonzero_hash<index_type>(size_[1]))
+    {}
+
+    /**
+     * Sets the matrix value at (row, col).
+     * If there is an existing value, it will be replaced by the sum of the
+     * existing and new value otherwise it will be.
```
Comment (tcojean) [inline include/ginkgo/core/base/matrix_assembly_data.hpp:128]: For completeness, maybe `@param` documentations?

File: include/ginkgo/core/base/matrix_assembly_data.hpp [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+     */
+    void add_value(index_type row, index_type col, value_type val)
+    {
+        auto ind = detail::symbolic_nonzero<index_type>{row, col};
+        nonzeros_[ind] += val;
+    }
+
+    /**
+     * Sets the matrix value at (row, col).
+     * If there is an existing value, it will be overwritten by the new value.
```
Comment (tcojean) [inline include/ginkgo/core/base/matrix_assembly_data.hpp:138]: Same, `@param`

File: include/ginkgo/core/base/matrix_assembly_data.hpp [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+     */
+    void set_value(index_type row, index_type col, value_type val)
+    {
+        auto ind = detail::symbolic_nonzero<index_type>{row, col};
+        nonzeros_[ind] = val;
+    }
+
+    /**
+     * Gets the matrix value at (row, col).
+     * Returns zero if it doesn't exist.
```
Comment (tcojean) [inline include/ginkgo/core/base/matrix_assembly_data.hpp:148]: `@param`

File: include/ginkgo/core/base/matrix_assembly_data.hpp [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+        auto it = nonzeros_.find(ind);
+        if (it == nonzeros_.end()) {
+            return zero<value_type>();
+        } else {
+            return it->second;
+        }
+    }
+
+    /**
+     * Returns true iff the matrix has an entry at (row, col).
```
Comment (tcojean) [inline include/ginkgo/core/base/matrix_assembly_data.hpp:162]: `@param`

File: include/ginkgo/core/base/matrix_assembly_data.hpp [hunk 5]
```diff
@@ -0,0 +1,224 @@
...
+     */
+    bool contains(index_type row, index_type col)
+    {
+        auto ind = detail::symbolic_nonzero<index_type>{row, col};
+        auto it = nonzeros_.find(ind);
+        return it != nonzeros_.end();
+    }
+
+    /** Returns the size of the matrix being assembled. */
```
Comment (tcojean) [inline include/ginkgo/core/base/matrix_assembly_data.hpp:171]: `@return`

File: include/ginkgo/core/base/matrix_assembly_data.hpp [hunk 6]
```diff
@@ -0,0 +1,224 @@
...
+
+    /** Returns the number of non-zeros in the (partially) assembled matrix. */
+    size_type get_num_stored_elements() const noexcept
+    {
+        return nonzeros_.size();
+    }
+
+    /**
+     * Returns a matrix_data instance containing the assembled non-zeros in
+     * row-major order to be used by all matrix formats.
```
Comment (tcojean) [inline include/ginkgo/core/base/matrix_assembly_data.hpp:182]: `@return`

**Review by tcojean — APPROVED** [review]
LGTM. Minor documentation comments.

**Review by pratikvn — APPROVED** [review]
LGTM!

File: include/ginkgo/core/base/matrix_assembly_data.hpp (hunk 3 above)
Comment (thoasm) [reply include/ginkgo/core/base/matrix_assembly_data.hpp:148]: And `@returns` here.

File: include/ginkgo/core/base/matrix_assembly_data.hpp [hunk 7]
```diff
@@ -0,0 +1,224 @@
...
+    }
+
+    /**
+     * Gets the matrix value at (row, col).
+     * Returns zero if it doesn't exist.
+     */
+    value_type get_value(index_type row, index_type col)
+    {
+        auto ind = detail::symbolic_nonzero<index_type>{row, col};
+        auto it = nonzeros_.find(ind);
```
Comment (thoasm) [inline include/ginkgo/core/base/matrix_assembly_data.hpp:153]: nit: both can be const here:
```suggestion
        const auto ind = detail::symbolic_nonzero<index_type>{row, col};
        const auto it = nonzeros_.find(ind);
```

File: include/ginkgo/core/base/matrix_assembly_data.hpp (hunk 4 above)
Comment (thoasm) [reply include/ginkgo/core/base/matrix_assembly_data.hpp:162]: And `@returns`

File: include/ginkgo/core/base/matrix_assembly_data.hpp [hunk 8]
```diff
@@ -0,0 +1,224 @@
...
+        return it != nonzeros_.end();
+    }
+
+    /** Returns the size of the matrix being assembled. */
+    dim<2> get_size() const noexcept { return size_; }
+
+    /** Returns the number of non-zeros in the (partially) assembled matrix. */
```
Comment (thoasm) [inline include/ginkgo/core/base/matrix_assembly_data.hpp:174]: `@returns` (or `@return`, they have the same meaning in Doxygen)

File: include/ginkgo/core/matrix/ell.hpp [hunk 9]
```diff
@@ -85,6 +85,7 @@ class Ell : public EnableLinOp<Ell<ValueType, IndexType>>,
     friend class Ell<to_complex<ValueType>, IndexType>;
 
 public:
+    using ReadableFromMatrixData<ValueType, IndexType>::read;
     using EnableLinOp<Ell>::convert_to;
     using EnableLinOp<Ell>::move_to;
```
Comment (thoasm) [inline include/ginkgo/core/matrix/ell.hpp:90]: nit: Previously, you placed it below `move_to` and `convert_to`:
```suggestion
    using EnableLinOp<Ell>::convert_to;
    using EnableLinOp<Ell>::move_to;
    using ReadableFromMatrixData<ValueType, IndexType>::read;
```

File: include/ginkgo/core/matrix/hybrid.hpp [hunk 10]
```diff
@@ -87,6 +87,7 @@ class Hybrid
 
 
 public:
+    using ReadableFromMatrixData<ValueType, IndexType>::read;
     using EnableLinOp<Hybrid>::convert_to;
     using EnableLinOp<Hybrid>::move_to;
```
Comment (thoasm) [inline include/ginkgo/core/matrix/hybrid.hpp:92]: Same here:
```suggestion
    using EnableLinOp<Hybrid>::convert_to;
    using EnableLinOp<Hybrid>::move_to;
    using ReadableFromMatrixData<ValueType, IndexType>::read;
```

File: include/ginkgo/core/matrix/sellp.hpp [hunk 11]
```diff
@@ -82,6 +82,7 @@ class Sellp : public EnableLinOp<Sellp<ValueType, IndexType>>,
     friend class Sellp<to_complex<ValueType>, IndexType>;
 
 public:
+    using ReadableFromMatrixData<ValueType, IndexType>::read;
     using EnableLinOp<Sellp>::convert_to;
     using EnableLinOp<Sellp>::move_to;
```
Comment (thoasm) [inline include/ginkgo/core/matrix/sellp.hpp:87]: same here:
```suggestion
    using EnableLinOp<Sellp>::convert_to;
    using EnableLinOp<Sellp>::move_to;
    using ReadableFromMatrixData<ValueType, IndexType>::read;
```

File: include/ginkgo/core/matrix/sparsity_csr.hpp [hunk 12]
```diff
@@ -79,6 +79,7 @@ class SparsityCsr
     friend class Csr<ValueType, IndexType>;
 
 public:
+    using ReadableFromMatrixData<ValueType, IndexType>::read;
     using EnableLinOp<SparsityCsr>::convert_to;
     using EnableLinOp<SparsityCsr>::move_to;
```
Comment (thoasm) [inline include/ginkgo/core/matrix/sparsity_csr.hpp:84]: and here:
```suggestion
    using EnableLinOp<SparsityCsr>::convert_to;
    using EnableLinOp<SparsityCsr>::move_to;
    using ReadableFromMatrixData<ValueType, IndexType>::read;
```

**Review by thoasm — APPROVED** [review]
LGTM, only minor nits.

File: include/ginkgo/core/base/matrix_assembly_data.hpp [hunk 13]
```diff
@@ -0,0 +1,224 @@
...
+#include <ginkgo/core/base/math.hpp>
+#include <ginkgo/core/base/matrix_data.hpp>
+#include <ginkgo/core/base/types.hpp>
+#include <ginkgo/core/base/utils.hpp>
+
+
+namespace gko {
+
+
+namespace detail {
```
Comment (yhmtsai) [inline include/ginkgo/core/base/matrix_assembly_data.hpp:54]:
```suggestion
namespace gko {
namespace detail {
```

File: include/ginkgo/core/base/matrix_assembly_data.hpp [hunk 14]
```diff
@@ -0,0 +1,224 @@
...
+    explicit symbolic_nonzero_hash(size_type num_cols) noexcept
+        : num_cols_{num_cols}
+    {}
+
+    std::size_t operator()(symbolic_nonzero<IndexType> nnz) const noexcept
+    {
+        return nnz.row * num_cols_ + nnz.column;
+    }
+
+    std::size_t num_cols_;
```
Comment (yhmtsai) [inline include/ginkgo/core/base/matrix_assembly_data.hpp:95]: Maybe also use `size_type` which is the same as the constructor input.

File: include/ginkgo/core/base/matrix_assembly_data.hpp [hunk 15]
```diff
@@ -0,0 +1,224 @@
...
+
+
+/**
+ * Type used to store indices used as keys in the unordered_map storing
+ * the nonzeros.
+ */
+template <typename IndexType>
+struct symbolic_nonzero {
+    using index_type = IndexType;
+    symbolic_nonzero() = default;
```
Comment (yhmtsai) [inline include/ginkgo/core/base/matrix_assembly_data.hpp:64]:
```suggestion

    symbolic_nonzero() = default;
```

**Review by yhmtsai — APPROVED** [review]
Can symbolic_nonzero be implemented by `std::pair`? or, will it require `make_pair` additionally, which is hard to use?

File: include/ginkgo/core/base/matrix_assembly_data.hpp [hunk 16]
```diff
@@ -0,0 +1,224 @@
...
+ * This structure is used as an intermediate type to assemble a sparse matrix.
+ *
+ * The matrix is stored as a set of nonzero elements, where each element is
+ * a triple of the form (row_index, column_index, value).
+ *
+ * New values can be added by using the matrix_assembly_data::add_entry or
+ * matrix_assembly_data::add_entries functions.
+ *
+ * @tparam ValueType  type of matrix values stored in the structure
+ * @tparam IndexType  type of matrix indexes stored in the structure
```
[the remark this answers is not in the document]
Comment (upsj) [reply include/ginkgo/core/base/matrix_assembly_data.hpp:112]: We're slightly inconsistent about this, even though I like indices more, too.

**Pull request merged**